### PR TITLE
CORE: skip zero size collectives

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Skip collective if size is zero

### CODE REVIEW

**Review by manjugv — APPROVED** [review]

File: src/core/ucc_lib.c [hunk 1]
```diff
@@ -353,6 +353,14 @@ ucc_status_t ucc_init_version(unsigned api_major_version,
         goto error;
     }
 
+    status = ucc_mpool_init(&lib->stub_tasks_mp, 0, sizeof(ucc_coll_task_t), 0,
+                            UCC_CACHE_LINE_SIZE, 8, UINT_MAX,
```
Comment (shimmybalsam) [inline src/core/ucc_lib.c:357]: do we need 8 elems_per_chunk for zero size messages?

File: src/core/ucc_coll.c [hunk 2]
```diff
@@ -297,18 +329,20 @@ UCC_CORE_PROFILE_FUNC(ucc_status_t, ucc_collective_post, (request),
     ucc_status_t status;
 
     if (ucc_global_config.coll_trace.log_level >= UCC_LOG_LEVEL_DEBUG) {
-        ucc_rank_t rank = task->team->params.team->rank;
-        if (ucc_global_config.coll_trace.log_level == UCC_LOG_LEVEL_DEBUG) {
-            if (rank == 0) {
+        if (task->team) {
+            ucc_rank_t rank = task->team->params.team->rank;
+            if (ucc_global_config.coll_trace.log_level == UCC_LOG_LEVEL_DEBUG) {
```
Comment (shimmybalsam) [inline src/core/ucc_coll.c:334]: perhaps remove second if (line 335) and else (line 340) and instead change to single if:
if (ucc_global_config.coll_trace.log_level != UCC_LOG_LEVEL_DEBUG || rank == 0) {
                ucc_log_component_collective_trace(
                    ucc_global_config.coll_trace.log_level,
                    "coll post: rank %d req %p, seq_num %u", rank, task,
                    task->seq_num);
}

File: src/core/ucc_coll.c [hunk 3]
```diff
@@ -388,18 +422,20 @@ UCC_CORE_PROFILE_FUNC(ucc_status_t, ucc_collective_finalize, (request),
     ucc_coll_task_t *task = ucc_derived_of(request, ucc_coll_task_t);
 
     if (ucc_global_config.coll_trace.log_level >= UCC_LOG_LEVEL_DEBUG) {
-        ucc_rank_t rank = task->team->params.team->rank;
-        if (ucc_global_config.coll_trace.log_level == UCC_LOG_LEVEL_DEBUG) {
-            if (rank == 0) {
+        if (task->team) {
+            ucc_rank_t rank = task->team->params.team->rank;
+            if (ucc_global_config.coll_trace.log_level == UCC_LOG_LEVEL_DEBUG) {
```
Comment (shimmybalsam) [inline src/core/ucc_coll.c:427]: perhaps remove second if (line 428) and else (line 433) and instead change to single if:
if (ucc_global_config.coll_trace.log_level != UCC_LOG_LEVEL_DEBUG || rank == 0) {
ucc_log_component_collective_trace(
ucc_global_config.coll_trace.log_level,
"coll post: rank %d req %p, seq_num %u", rank, task,
task->seq_num);
}

**Review by shimmybalsam — APPROVED** [review]

File: src/core/ucc_lib.c (hunk 1 above)
Comment (Sergei-Lebedev) [reply src/core/ucc_lib.c:360]: 8 is enough and it doesn't consume much memory

File: src/core/ucc_coll.c (hunk 2 above)
Comment (Sergei-Lebedev) [reply src/core/ucc_coll.c:334]: it will be different in that case since rank 0 will print line twice

**Review by samnordmann — APPROVED** [review]

File: src/core/ucc_coll.c [hunk 4]
```diff
@@ -157,6 +157,15 @@ static ucc_status_t ucc_coll_args_check_mem_type(ucc_coll_args_t *coll_args,
     };
 }
 
+#define UCC_COLL_TYPE_SKIP_ZERO_SIZE \
```
[the remark this answers is not in the document]
Comment (Sergei-Lebedev) [reply src/core/ucc_coll.c:160]: we have a particular application that needs to skip those collectives, reduce scatter is not needed for this application. We can of course add others but it's a question of semantic, for example should zero size reduce_scatter work as barrier or not

**Comment by Sergei-Lebedev** [comment]
bot:retest

**Pull request merged**